elastic/logstash
# metrics: live-timers

## Release notes


Improves timer metrics to include in-progress execution, so that the effects of blocking operations are apparent _while_ they are blocking instead of only _after_ they are resolved.


## What does this PR do?

Introduces a first-class TimerMetric, and provides an implementation that is capable of tracking live concurrent execution _before_ that execution has completed.

## Why is it important/What is the impact to the user?

Incrementing timers after execution effectively makes it so that metrics tracking blocking operations _cannot_ accurately report how much time is being spent. Here we first introduce the timer as a first-class metric, and add an implementation that can calculate cumulative elapsed time using a checkpoint along with a wall-clock delta and a known-constant concurrency since that checkpoint.

This allows metrics for blocked or stuck operations to show just how much resources they are taking, instead of reporting zeroes while stuck. This is especially useful now that we are exposing "flow" metrics that show rate-of change; if a pipeline's workers are _all_ busy with long-running operations, we want to see our new `worker_concurrency` flow metric's `current` value be high and not zero.

## Checklist




## Author's Checklist



## How to test this PR locally

Run a pipeline that blocks the workers during execution, and observe the node stats APIs flow metrics. Without this PR, our `current` `worker_concurrency` is frequently 0, but with it we get a more accurate view that we are using up all available resources.

```
bin/logstash -e 'input { generator {} } filter { ruby { code => "sleep(1)"} } output { sink {} }'
```
